### PR TITLE
[GLUTEN-3364][VL] SplitPreloadPerDriver (SplitPreloadPerDriver = true, IOThreads > 0) causes unexpected crashes

### PR DESCRIPTION
**Pull request by zhztheplayer**
This fixes issue #3364.

~~After applying the patch, global memory pool will be used when IO threads > 0 for memory allocations in Hive connector. However side-effect that the allocations will be out-of-track by Spark memory manager will be brought.~~

Update: Switching to global memory pool doesn't help since the async executor holds task's memory pool anyway. Added a wait strategy (up to 1.5s) until async allocations are released. Some minor cleanups to connector initialization procedure.

### CODE REVIEW

**Comment by zhouyuan** [comment]
maybe we could enable multiple threads pool by default, to make the code to be tested in more cases

File: cpp/velox/compute/VeloxBackend.cc [hunk 1]
```diff
@@ -342,19 +280,130 @@ void VeloxBackend::initCache(const std::unordered_map<std::string, std::string>&
...
+  auto properties = std::make_shared<const velox::core::MemConfig>(configurationValues);
+
+  if (ioThreads > 0) {
+    ioExecutor_ = std::make_unique<folly::IOThreadPoolExecutor>(ioThreads);
+    FLAGS_split_preload_per_driver = splitPreloadPerDriver;
+    // Use global memory pool for hive connector if SplitPreloadPerDriver was enabled. Otherwise, the allocated memory
+    // blocks might cause unexpected behavior (e.g. crash) since the allocations were proceed in background IO threads.
+    velox::connector::registerConnector(std::make_shared<MemoryPoolReplacedHiveConnector>(
+        gluten::defaultLeafVeloxMemoryPool().get(), kHiveConnectorId, properties, ioExecutor_.get()));
+  } else {
```
Comment (Yohahaha) [inline cpp/velox/compute/VeloxBackend.cc:403]: I prefer use global memory pool only when split preload enabled. IO executor does not always has problem.

**Comment by zhouyuan** [comment]
@zhztheplayer can you please do a rebase?

**Comment by zhztheplayer** [comment]
> @zhztheplayer can you please do a rebase?

done

**Comment by zhouyuan** [comment]
I verified locally with io_threads = 16, tpcds workload passed.

**Review by zhouyuan — APPROVED** [review]
👍

**Comment by zhouyuan** [comment]
@Yohahaha @zhli1142015 This patch should fix the memory pool issue when enabling io thread pool.

**Comment by zhli1142015** [comment]
> @Yohahaha @zhli1142015 This patch should fix the memory pool issue when enabling io thread pool.

Thanks.

**Comment by zhli1142015** [comment]
Hello @zhouyuan and @zhztheplayer ,
We still observed some random issues with this fix, could you help check?
Thanks.
```

E20231129 15:53:18.857537 26213 Exceptions.h:69] Line: /var/git/Velox/velox/common/memory/Memory.cpp:80, Function:~MemoryManager, Expression: numPools() == 0 (1 vs. 0) There are 1 unexpected alive memory pools allocated by user on memory manager destruction:
Memory Manager[capacity UNLIMITED alignment 64B usedBytes 6.73MB number of pools 1
List of root pools:
	__default_root__
	WholeStageIterator_root
Memory Allocator[MALLOC capacity UNLIMITED allocated bytes 7059648 allocated pages 1536 mapped pages 0]
ARBITRATOR[GLUTEN] CAPACITY 8388608.00TB STATS[numRequests 0 numSucceeded 0 numAborted 0 numFailures 0 numNonReclaimableAttempts 0 numReserves 0 numReleases 0 queueTime 0us arbitrationTime 0us reclaimTime 0us shrunkMemory 0B reclaimedMemory 0B maxCapacity 0B freeCapacity 0B]], Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (1 vs. 0) There are 1 unexpected alive memory pools allocated by user on memory manager destruction:
Memory Manager[capacity UNLIMITED alignment 64B usedBytes 6.73MB number of pools 1
List of root pools:
	__default_root__
	WholeStageIterator_root
Memory Allocator[MALLOC capacity UNLIMITED allocated bytes 7059648 allocated pages 1536 mapped pages 0]
ARBITRATOR[GLUTEN] CAPACITY 8388608.00TB STATS[numRequests 0 numSucceeded 0 numAborted 0 numFailures 0 numNonReclaimableAttempts 0 numReserves 0 numReleases 0 queueTime 0us arbitrationTime 0us reclaimTime 0us shrunkMemory 0B reclaimedMemory 0B maxCapacity 0B freeCapacity 0B]]
Retriable: False
Expression: numPools() == 0
Function: ~MemoryManager
File: /var/git/Velox/velox/common/memory/Memory.cpp
Line: 80
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 3  _ZN8facebook5velox6memory13MemoryManagerD1Ev
# 4  _ZN6gluten18VeloxMemoryManagerD1Ev
# 5  _ZN6gluten18VeloxMemoryManagerD0Ev
# 6  0x00007f14b0fc8607

```

**Comment by Yohahaha** [comment]
@zhli1142015 does remove taskHolder work？

**Comment by Yohahaha** [comment]
seems VMM dtor order was wrong, mem pool should be released first.

**Comment by zhztheplayer** [comment]
Hi @zhli1142015, would you like to search for the following log prior to the log your pasted?

```
There are still outstanding Velox memory allocations. Waiting for ... ms to let possible async tasks done...
```

The log was added by this PR. Since `~VeloxMemoryManager` should check for unclosed memory pools.

**Comment by zhli1142015** [comment]
> @zhli1142015 does remove taskHolder work？

In our test, if there are errors happen in IO thread pool threads, for example in our connector, then executor would crash directly. So removing the `taskHolder` doesn't solve all issues. By now we revert that change.
Thanks.

**Comment by zhli1142015** [comment]
> Hi @zhli1142015, would you like to search for the following log prior to the log your pasted?
> 
> ```
> There are still outstanding Velox memory allocations. Waiting for ... ms to let possible async tasks done...
> ```
> 
> The log was added by this PR. Since `~VeloxMemoryManager` should check for unclosed memory pools.

Looks now only warn and error native logs would be outputted, let me change the log settings and try to reproduce the errors.
Thanks.

**Comment by zhli1142015** [comment]
@zhztheplayer , thanks.
```
I20231205 06:12:10.166657 22564 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 50 ms to let possible async tasks done... 
I20231205 06:12:10.192996 22568 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_84_TID_3509 after running for 1789 ms.
I20231205 06:12:10.193049 22568 Task.cpp:1742] Terminating task Gluten_Stage_84_TID_3509 with state Finished after running for 1789 ms.
I20231205 06:12:10.216785 22564 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 100 ms to let possible async tasks done... 
I20231205 06:12:10.316962 22564 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 200 ms to let possible async tasks done... 
I20231205 06:12:10.517093 22564 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 400 ms to let possible async tasks done... 
I20231205 06:12:10.519326 22557 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_84_TID_3472 after running for 2989 ms.
I20231205 06:12:10.519361 22557 Task.cpp:1742] Terminating task Gluten_Stage_84_TID_3472 with state Finished after running for 2989 ms.
I20231205 06:12:10.583216 22570 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_84_TID_3503 after running for 2340 ms.
I20231205 06:12:10.583281 22570 Task.cpp:1742] Terminating task Gluten_Stage_84_TID_3503 with state Finished after running for 2340 ms.
I20231205 06:12:10.609946 22565 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_84_TID_3510 after running for 2164 ms.
I20231205 06:12:10.609987 22565 Task.cpp:1742] Terminating task Gluten_Stage_84_TID_3510 with state Finished after running for 2164 ms.
I20231205 06:12:10.917227 22564 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 800 ms to let possible async tasks done... 
I20231205 06:12:11.458278 22566 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_88_TID_3633 after running for 2431 ms.
I20231205 06:12:11.458331 22566 Task.cpp:1742] Terminating task Gluten_Stage_88_TID_3633 with state Finished after running for 2431 ms.
I20231205 06:12:11.512677 22569 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_84_TID_3476 after running for 3750 ms.
I20231205 06:12:11.512735 22569 Task.cpp:1742] Terminating task Gluten_Stage_84_TID_3476 with state Finished after running for 3750 ms.
I20231205 06:12:11.670019 22563 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_88_TID_3622 after running for 2654 ms.
I20231205 06:12:11.670085 22563 Task.cpp:1742] Terminating task Gluten_Stage_88_TID_3622 with state Finished after running for 2654 ms.
I20231205 06:12:11.724128 22565 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 50 ms to let possible async tasks done... 
I20231205 06:12:11.774241 22565 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 100 ms to let possible async tasks done... 
I20231205 06:12:11.874421 22565 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 200 ms to let possible async tasks done... 
I20231205 06:12:11.893692 22560 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_88_TID_3646 after running for 2616 ms.
I20231205 06:12:11.893734 22560 Task.cpp:1742] Terminating task Gluten_Stage_88_TID_3646 with state Finished after running for 2616 ms.
I20231205 06:12:11.966938 22555 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_88_TID_3643 after running for 2730 ms.
I20231205 06:12:11.966985 22555 Task.cpp:1742] Terminating task Gluten_Stage_88_TID_3643 with state Finished after running for 2730 ms.
I20231205 06:12:11.998804 22559 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_88_TID_3663 after running for 2201 ms.
I20231205 06:12:11.998848 22559 Task.cpp:1742] Terminating task Gluten_Stage_88_TID_3663 with state Finished after running for 2201 ms.
I20231205 06:12:12.074555 22565 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 400 ms to let possible async tasks done... 
I20231205 06:12:12.280316 22562 Task.cpp:1055] All drivers (1) finished for task Gluten_Stage_87_TID_3582 after running for 3406 ms.
I20231205 06:12:12.280360 22562 Task.cpp:1742] Terminating task Gluten_Stage_87_TID_3582 with state Finished after running for 3406 ms.
I20231205 06:12:12.474665 22565 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 800 ms to let possible async tasks done... 
E20231205 06:12:13.274852 22565 Exceptions.h:69] Line: /var/git/Velox/velox/common/memory/Memory.cpp:80, Function:~MemoryManager, Expression: numPools() == 0 (1 vs. 0) There are 1 unexpected alive memory pools allocated by user on memory manager destruction:
Memory Manager[capacity UNLIMITED alignment 64B usedBytes 278.40MB number of pools 1
List of root pools:
	__default_root__
	WholeStageIterator_root
Memory Allocator[MALLOC capacity UNLIMITED allocated bytes 291924480 allocated pages 63808 mapped pages 63488]
ARBITRATOR[GLUTEN] CAPACITY 8388608.00TB STATS[numRequests 0 numSucceeded 0 numAborted 0 numFailures 0 numNonReclaimableAttempts 0 numReserves 0 numReleases 0 queueTime 0us arbitrationTime 0us reclaimTime 0us shrunkMemory 0B reclaimedMemory 0B maxCapacity 0B freeCapacity 0B]], Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (1 vs. 0) There are 1 unexpected alive memory pools allocated by user on memory manager destruction:
Memory Manager[capacity UNLIMITED alignment 64B usedBytes 278.40MB number of pools 1
List of root pools:
	__default_root__
	WholeStageIterator_root
Memory Allocator[MALLOC capacity UNLIMITED allocated bytes 291924480 allocated pages 63808 mapped pages 63488]
ARBITRATOR[GLUTEN] CAPACITY 8388608.00TB STATS[numRequests 0 numSucceeded 0 numAborted 0 numFailures 0 numNonReclaimableAttempts 0 numReserves 0 numReleases 0 queueTime 0us arbitrationTime 0us reclaimTime 0us shrunkMemory 0B reclaimedMemory 0B maxCapacity 0B freeCapacity 0B]]
Retriable: False
Expression: numPools() == 0
Function: ~MemoryManager
File: /var/git/Velox/velox/common/memory/Memory.cpp
Line: 80
Stack trace:
# 0  _ZN8facebook5velox7process10StackTraceC1Ei
# 1  _ZN8facebook5velox14VeloxExceptionC1EPKcmS3_St17basic_string_viewIcSt11char_traitsIcEES7_S7_S7_bNS1_4TypeES7_
# 2  _ZN8facebook5velox6detail14veloxCheckFailINS0_17VeloxRuntimeErrorERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEvRKNS1_18VeloxCheckFailArgsET0_
# 3  _ZN8facebook5velox6memory13MemoryManagerD1Ev
# 4  _ZN6gluten18VeloxMemoryManagerD1Ev
# 5  _ZN6gluten18VeloxMemoryManagerD0Ev
# 6  0x00007f46d74f7df1

```

**Comment by zhztheplayer** [comment]
Noticed this line 

```
I20231205 06:12:12.474665 22565 VeloxMemoryManager.cc:293] There are still outstanding Velox memory allocations. Waiting for 800 ms to let possible async tasks done... 
```

This means after (100 + 200 + 400 + 800) ms = 1.5s the memory used by split preloading of the task still not gets released. Which means split preloading may costs >1.5s. Do you have some clues why preloading took this long time?

**Pull request merged**